src/integTest/kotlin/com/ampnet/payoutservice/blockchain/Web3jBlockchainServiceIntegTest.kt

```diff
@@ -572,6 +572,46 @@ class Web3jBlockchainServiceIntegTest : TestBase() {
         }
     }
 
+    @Test
+    fun mustCorrectlyFindContractDeploymentBlockNumber() {
+        val mainAccount = accounts[0]
+
+        suppose("100 blocks will be mined") {
+            hardhatContainer.mineUntil(maxAttempts = 100) { false }
+        }
+
+        val contract = suppose("simple ERC20 contract is deployed") {
+            SimpleERC20.deploy(
+                hardhatContainer.web3j,
+                mainAccount,
+                DefaultGasProvider(),
+                listOf(mainAccount.address),
+                listOf(BigInteger("10000")),
+                mainAccount.address
+            ).sendAndMine()
+        }
+
+        val deploymentBlockNumber = contract.transactionReceipt.get().blockNumber
+
+        suppose("250 blocks will be mined") {
+            hardhatContainer.mineUntil(maxAttempts = 250) { false }
+        }
+
+        verify("correct contract deployment block number is returned") {
+            val service = Web3jBlockchainService(hardhatProperties())
+            val foundDeploymentBlock = service.findContractDeploymentBlockNumber(
+                chainId = Chain.HARDHAT_TESTNET.id,
+                contractAddress = ContractAddress(contract.contractAddress)
+            )
+
+            // it's okay to have either the exact deployment block number or one block BEFORE deployment, but not after
```

Comment:
Is it possible to have a block number of deplyoment before deployment?

Reply:
I'm, not sure what you are asking, but here is what I mean:
In binary search, if we get to something like range `[400, 401]` and `401` is our deployment block, we will return `400` because the range is 1 or less and we always return the lower number. That means we will have one block before our contract was deployed instead of the actual block, but that would be fine since its close enough that we won't miss any contract calls - we just have an extra block with no contract interactions yet.